make slugs behave more like GitHub

Flet opened the pull request:
Thanks for sharing this package, its quite useful! :)

## Problem

I noticed that `marky-markdown` is not consistent with GitHub when generating anchor IDs. Specifically, its filtering out underscores and trailing dashes, which are left alone on GitHub.

Here is what I saw coming out of `marky-markdown`:
```
> var md = require('marky-markdown')
> md('## test(\n## test)\n## test_\n## test-\n## test=\n').html()
```
```html
<h2 id="user-content-test" class="deep-link"><a href="#test">test(</a></h2>
<h2 id="user-content-test" class="deep-link"><a href="#test">test)</a></h2>
<h2 id="user-content-test" class="deep-link"><a href="#test">test_</a></h2>
<h2 id="user-content-test" class="deep-link"><a href="#test">test-</a></h2>
<h2 id="user-content-test" class="deep-link"><a href="#test">test=</a></h2>
```

...and here is an example of what GitHub does (pulling from this this [gist](https://gist.github.com/Flet/1f46a2af2f6b3bf378cf)):
```html
<h2><a id="user-content-test-14" class="anchor" href="#test-14" aria-hidden="true"><span class="octicon octicon-link"></span></a>test(</h2>
<h2><a id="user-content-test-15" class="anchor" href="#test-15" aria-hidden="true"><span class="octicon octicon-link"></span></a>test)</h2>
<h2><a id="user-content-test_" class="anchor" href="#test_" aria-hidden="true"><span class="octicon octicon-link"></span></a>test_</h2>
<h2><a id="user-content-test-" class="anchor" href="#test-" aria-hidden="true"><span class="octicon octicon-link"></span></a>test-</h2>
<h2><a id="user-content-test-16" class="anchor" href="#test-16" aria-hidden="true"><span class="octicon octicon-link"></span></a>test=</h2>
```

`marky-markdown` relies on `slugg` for which deviates from [GitHub's punctuation filtering behavior](https://github.com/jch/html-pipeline/blob/master/lib/html/pipeline/toc_filter.rb#L35).

The other problem is that some IDs become duplicates after being filtered. `marky-markdown` made all the IDs `test`, but GitHub will add a suffix to ensure they are unique ('test', 'test-2', 'test-3').

## Solution
This PR switches from `slugg` to [`slugger-unique`](https://github.com/flet/slugger-unique), which uses [`slugger`](https://github.com/HenrikJoreteg/slugger) to generate unique IDs. `slugger-unique` takes it further by adding a suffix for duplicate IDs, mimicking GitHub's behavior.

Tests have also been added to cover the cases mentioned.

wooorm commented:
Cool!

zeke commented:
:+1:

Qard commented:
It's close to github style, but not quite there. Have a look at the links in https://www.npmjs.com/package/amqp for a good test case. All the `[, callback]` bits produce `--` in github, but this reduces it to `-`.

Flet commented:
Thanks for the feedback, @Qard!

Looks like `slugger` did not quite cut it, so I built [github-slugger](https://github.com/Flet/github-slugger) to fix the test case you pointed out.

I added a test to cover @Quard's case and squashed everything to a single commit.

Flet commented:
Wow, lots of activity on this package! :+1: I rebased this PR to get it caught up. Please let me know if there are any other questions/concerns!

ashleygwilliams commented:
this looks great to me. i'd like to get a member of our web team to review it and then pending their approval, i'll look into merging and cutting a new release!

cc @aredridel @rockbot @jefflembeck would one/all of you take a quick look and give it a :+1: ?

Flet commented:
Thanks for considering!

I noticed the "needs tests" label was added. Are the included tests not sufficient? I'm happy to add more if required!

ashleygwilliams commented:
whoops! i missed em. doing issue triage too early. the tests look great. waiting to hear back from the web team, but i'd like to add this to 5.6

zeke commented:
slugs!!!

:snail:

Flet commented:
Thanks! Exciting! :tada:
